[Inductor][FX passes] Pre grad batch linear LHS fusion

yanboliang opened the pull request:
This is a popular pattern in many internal user cases, we have two versions (pre and post grad) and found the pre grad version has more perf gain, which makes sense in theory as this corresponding backward graph doesn't have this pattern. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

jackiexu1992 commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -120,19 +124,98 @@ def fuse(self, graph, subset):
+            batch_weights.append(weight)
+            if bias:
+                batch_biases.append(bias)
+            split_sections.append(weight.meta["example_value"].shape[0])
```
Noob question. Somehow I still saw this warning in my local run. "[2023-08-02 20:11:05,605] torch._inductor.fx_passes.split_cat: [WARNING] example value absent for node: split_15"

yanboliang replied:
I saw it as well, but it's from the split cat pass. Some nodes don't have example value and we won't apply split-cat patterns. We should look at why the example value absent in the input graph, which should be a dynamo issue IMO. I'll follow up this in a separate thread.

yanboliang committed:
[Inductor][FX passes] Pre grad batch linear LHS fusion (#106497)

Summary:
This is a popular pattern in many internal user cases, we have two versions (pre and post grad) and found the pre grad version has more perf gain, which makes sense in theory as this corresponding backward graph doesn't have this pattern.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106497

Reviewed By: jackiexu1992

Differential Revision: D48016423

Pulled By: yanboliang

fbshipit-source-id: 1674bb037fe55eba64ae1f2850e3554bf7cb7340
